### DISCUSSION TITLE
How to use expressions in subcaptions

### DISCUSSION THREAD

**Discussion by dmkaplan2000** [quarto-dev/quarto-cli#7308] [Q&A]
### Description

I want to use expressions in subcaptions, but I can't seem to figure out how to do it. For example, when I render the following document:

````qmd
---
format: html
---

```{r}
x = "A test label"
```

```{r}
#| label: fig-test
#| fig-cap: !expr x
#| fig-subcap:
#|   - !expr x
#|   - !expr x
plot(1:10)
plot(1:20)
```
````

The overall figure caption appears as it should ("A test label"), but the captions for the subfigures are not correct, appearing as "expression(x)". Is there a way to fix this? I have tried various forms of quoting around `x`, but nothing seems to work.

**Comment by mcanouil** [comment]
Which version are you using?

Note that `!expr` is not recommended (even if is is supported) in Quarto.
With 1.4 you can do the following:

````qmd
---
format: html
---

```{r}
x <- "A test label"
```

:::: {#fig-test}

::: {#fig-test-1}
```{r}
plot(1:10)
```

`{r} x`
:::

::: {#fig-test-2}
```{r}
plot(1:10)
```

`{r} x`
:::

`{r} x`
::::
````

<img width="573" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/d1e64053-db28-44fd-b8ff-1c92210b3eb7">

**Reply by mcanouil** [reply]
@cderv is `!expr` even supported in `fig-subcap`?

**Reply by cscheid** [reply]
(In passing, you'll note that the source code for these subfigures is currently being included in the figure itself. That's a quarto bug that we'll resolve before 1.4 is out.)

**Reply by harrelfe** [reply]
In other contexts the notation for R expressions in text is backtick r expression backtick without putting r in braces.  Why not use that notation for consistency?

**Reply by mcanouil** [reply]
It's the knitr syntax which is unrelated to consistency as far as Quarto is concerned.
Quarto brings inline expression beyond knitr.
See <https://quarto.org/docs/prerelease/1.4/inline.html>.

Also knitr in my opinion is not consistent `` ```r `` is not evaluated but `` ```{r} `` is so why `` `r `` also is?

Quarto makes this consistent for R but also for other languages.
Computed code cells and inline expressions use brackets.

**Comment by dmkaplan2000** [comment]
Huh, OK, I have used that approach for external images, but never for figures that are output from R. Are `fig-cap` and `fig-subcap` being deprecated then? We could essentially eliminate them and generate all figures and tables the way you are proposing.

To answer @cderv, `!expr` does do things for `fig-subcap`, but not at all what you expect. It seems to wrap the argument in `expression()`.

I am using Quarto 1.4.429.

**Reply by cscheid** [reply]
> Huh, OK, I have used that approach for external images, but never for figures that are output from R. Are fig-cap and fig-subcap being deprecated then?

They're not deprecated, but !expr is going to become less supported as time goes by, simply because it's a knitr-specific hack that makes a lot of quarto features impossible to implement correctly.

We're encouraging people to use the div syntax because that makes it particularly easy to write captions that include "fancy" markdown, including crossreferences, shortcodes, and computational content. Those are hard to do correctly in `!expr` in all settings.

**Reply by dmkaplan2000** [reply]
Yes, https://github.com/quarto-dev/quarto-cli/issues/1959 was my question way back on the issue of fancy caption. This is a great fix.

**Comment by dmkaplan2000** [comment]
Also, is `!expr` not recommended for all chunk arguments or just things like figure captions? For example, I often include things like:

````qmd
#| eval: !expr file.exists("my-results.RDS")
````

If there a different approach to achieving this result?

**Reply by mcanouil** [reply]
It is because it's specific to `knitr` and does not work in Jupyter.
Also not all options supports this, for instance even in 1.3.450, what you are trying does not work so it was not broken by the new cross-reference system.

`eval` / `include` has no alternatives for conditional evaluation (there is an issue open on the matter but more general than `knitr`).

**Reply by cscheid** [reply]
The particular `eval: !expr` use case is currently one that quarto doesn't offer a great alternative for (because the evaluation string stops the entire cell from executing). So if you find that one useful, you'll probably need to continue doing so. We have no plans on forbidding the usage (it's actually hard for us to do it)

But you should note that `!expr` makes it impossible, for example, for us to so YAML validation. So you could get a strange knitr or downstream quarto error that we currently can't protect against.

**Reply by dmkaplan2000** [reply]
Thanks for the explanation and will keep this in mind. `eval: !expr` is really helpful for making code that skips over super slow analyses if they are already available, so hopefully it will either be kept or an alternative will be found. How is that sort of thing done in python if `!expr` doesn't work?

**Reply by mcanouil** [reply]
> eval: !expr is really helpful for making code that skips over super slow analyses if they are already available

For this exact use case, there is `targets` which actually does more that just checking if it exists, but also checks if something changed in the step before to ensure integrity of the workflow.

**Reply by cderv** [reply]
> How is that sort of thing done in python if !expr doesn't work?

I think there is no good alternative for this in Jupyter. You can't tweak eval conditionally, but we introduce `when-format` conditional on div to allow to hide from output. (execution still happens though). 

> Also, is !expr not recommended for all chunk arguments or just things like figure captions?

To be clear, `!expr` still works, and will still be working with **knitr**. it is just not recommanded for the several reason shared above about specific feature (like Jupyter support, YAML validation), and because it won't work on all YAML options (can work only in chunks for example, not yaml header). 

So overall, it is best to avoid when possible, especially when teaching, or working on cross language project. 

However, like other **knitr** specific feature, this is helpful sometimes. 

> is !expr even supported in fig-subcap?

It is working, but here you are confronted exactly on why `!expr` is not recommended (or at least considered advanced usage. 

This is not valid `!expr` usage 
```yaml
#| fig-subcap:
#|   - !expr x
#|   - !expr x
````

This is valid `!expr` usage 
````yaml
#| fig-subcap: !expr c(x, x)
````

For the context, `!expr` is a specific tag feature from **yaml** R package that will evaluate its value to get YAML structure. So if you want a yaml list for `subcap`, you need to use R expression that does it. 

Here are some example to understand. This is how the YAML list syntax is translated
``` r
res <- yaml::yaml.load("fig-subcap:\n  - x\n  - x")
str(res)
#> List of 1
#>  $ fig-subcap: chr [1:2] "x" "x"
res$`fig-subcap`
#> [1] "x" "x"
```

So if you provide correctly with a vector with `!expr` and `eval.expr` activated (which knitr does by default)
``` r
res <- yaml::yaml.load("fig-subcap: !expr c('x', 'x')", eval.expr = TRUE)
str(res)
#> List of 1
#>  $ fig-subcap: chr [1:2] "x" "x"
res$`fig-subcap`
#> [1] "x" "x"
```

I hope it helps clarify how this works.

**Reply by cscheid** [reply]
> How is that sort of thing done in python if `!expr` doesn't work?

Conditionals, like you can also do in `{r}`:

````
```{r}
#| eval: !expr foo
code
```
````

is fairly close to

````
```{r}
if (foo) {
  code
}
```
````

**Reply by mcanouil** [reply]
Not if you want to do something like:

````qmd
---
title: "Quarto Playground"
format: html
execute:
  echo: false
---

This is a playground for Quarto.

```{r}
#| classes: callout-note
#| output: asis
if (FALSE) {
  cat("1 + 1")
}
```
````

This will produce an empty callout.

<img width="575" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/1d8738e0-f748-4a6a-ae1d-157bed77cd6f">

~~Edit: oh it seems it no longer output an empty callout.~~

Note that Jupyter/Python does it right for this example but not the same way.
classes/output options behave completely differently between knitr/jupyter (kind of related to the inline differences between the two).

````qmd
---
title: "Quarto Playground"
format: html
execute:
  echo: false
---

This is a playground for Quarto.

```{python}
#| classes: callout-note
#| output: asis
from IPython.display import display, Markdown
if True:
  display("*1 + 1*")
```

```{python}
#| classes: callout-note
from IPython.display import display, Markdown
if True:
  display("*1 + 1*")
```

```{python}
#| classes: callout-note
from IPython.display import display, Markdown
if False:
  display("*1 + 1*")
```
````

<img width="524" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/5f7a1ea9-1b97-457a-89aa-b79028b9218a">

**Reply by cscheid** [reply]
> Not if you want to do something like:

We never promised support for that. It's remarkable that it works, but please, let's not encourage folks to use this :)

**Reply by mcanouil** [reply]
That's kind of documented though.

**Reply by cderv** [reply]
`eval: false` is different than `if(FALSE)` in the sense that knitr will really skip evaluation of code. And when conditionally using a chunk, using `if` syntax when `echo: true` will show you source code like you don't really want. That is why it is an option and not just à `if` in the code chunk. 

Anyhow, this is a knitr feature that still works well. One just need to know how to use it correctly in chunk and know the limitation.
